packages/react-hooks/src/useDebouncedValue.ts

```diff
@@ -1,25 +1,44 @@
-import { useEffect, useState } from 'react';
+import { useEffect, useMemo, useState } from 'react';
 
 /**
  * Debounces a value.
  * Returns the initial value immediately.
  * Returns the latest value after no changes have occurred for the debounce duration.
  * @param value Value to debounce
  * @param debounceMs Amount of time to debounce
- * @returns The debounced value
+ * @returns The debounced value + whether the value is still debouncing
  */
-export function useDebouncedValue<T>(value: T, debounceMs: number): T {
+export function useDebouncedValue<T>(
+  value: T,
+  debounceMs: number
+): { isDebouncing: boolean; value: T } {
+  const [isDebouncing, setIsDebouncing] = useState(true);
   const [debouncedValue, setDebouncedValue] = useState<T>(value);
+
+  // Set isDebouncing to true immediately whenever the value changes. Using
+  // `useMemo` instead of `useEffect` so that state is never out of sync whenever
+  // value and / or debounceMs have changed.
+  useMemo(() => {
```

Comment:
An alternative to this would be 

```
  const [previousValue, setPreviousValue] = useState(value);
  const [previousDebounceMs, setPreviousDebounceMs] = useState(debounceMs);

  if (value !== previousValue || debounceMs !== previousDebounceMs) {
    setIsDebouncing(true);
    setPreviousValue(value);
    setPreviousDebounceMs(debounceMs);
  }
  ```
  
  This actually aligns with an example in the React docs whereas not sure if setting state inside of `useMemo` is recommended.

Reply:
Yeah, I didn't know this pattern until today.